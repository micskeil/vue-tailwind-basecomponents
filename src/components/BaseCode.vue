<template>
  <InputContainer element-type="code">
    <div class="code-tools">
      <div class="copy">
        <span v-if="copied" class="copy__confirmation">Copied!</span>
        <BaseIcon class="copy__icon" icon="content-copy" :size="16" @click="copyCode" />
      </div>
    </div>
    <pre class="code"><code
:class="`language-${language}`"
                            v-html="html"
    /></pre>
    <div class="code-language">
      {{ language }}
    </div>
  </InputContainer>
</template>

<script setup lang="ts">
  import InputContainer from './InputContainer.vue';
  import Prism from 'prismjs';
  import 'prismjs/themes/prism-tomorrow.css';
  import { ref } from 'vue';

  const props = defineProps({
    /**
     * The code to display
     */
    code: {
      type: String,
      required: true,
    },
    /**
     * The language to use for syntax highlighting,
     * check PrismJS docs for supported languages
     * @default javascript
     * TODO: Add better typing
     */
    language: {
      type: String,
      default: 'javascript',
    },
    // TODO: Add support for different styles
  });

  // Returns a highlighted HTML string
  const html = Prism.highlight(props.code, Prism.languages.javascript, 'javascript');

  const copied = ref(false);
  function copyCode() {
    copied.value = true;
    navigator.clipboard.writeText(props.code);
  }
</script>

<style lang="postcss" scoped>
  .code-container {
    @apply relative overflow-hidden rounded-xl border-none bg-gray-800;

    .code {
      @apply overflow-x-auto p-2 pb-4 pt-8 lg:p-4 lg:pb-8 lg:pt-10;
    }

    .code-tools {
      @apply absolute right-0 top-0 pr-2 pt-2 lg:pr-4;

      .copy {
        @apply flex items-center gap-1 rounded-b text-xs text-gray-500;
        &:deep(.icon) {
          @apply mr-1 inline-block cursor-pointer fill-current p-1 text-gray-200 transition-transform
            duration-300 ease-in-out hover:text-gray-600;
          &:hover {
            @apply rotate-12;
          }
          &:active {
            @apply scale-125;
          }
        }
      }
    }

    .code-language {
      @apply absolute left-0 top-0 rounded-b pl-2 pt-2 text-xs text-gray-500 lg:pl-4;
    }
  }
</style>
