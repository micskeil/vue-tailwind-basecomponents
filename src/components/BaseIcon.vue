<template>
  <span :title="label" class="icon" role="img" :aria-label="label" :aria-labelledby="parentId">
    <svg class="icon__native" viewBox="0 0 24 24" :width="width" :class="[`fill-${color}`, `icon-${icon}`]">
      <path :d="path" />
    </svg>
  </span>
</template>

<script setup lang="ts">
  // import the mdi library and annotate its type
  import * as mdi from '@mdi/js';
  import { computed } from 'vue';

  interface MdiIcons {
    [key: string]: string;
  }
  const mdiIcons: MdiIcons = mdi;

  const props = defineProps({
    label: {
      type: String,
      default: undefined,
    },
    icon: {
      type: String,
      required: true,
    },
    color: {
      type: String,
      default: 'primary',
    },
    size: {
      type: [Number, String],
      default: 24,
    },
    parentId: {
      type: [String],
      default: undefined,
    },
  });

  /**
   * Get the width of the icon.
   */
  const width = computed(() => {
    if (typeof props.size === 'number') {
      return props.size;
    }
    // turn tailwind text-sizes into numbers

    return {
      xs: 12,
      sm: 16,
      base: 24,
      lg: 32,
      xl: 48,
      '2xl': 64,
      '3xl': 96,
      '4xl': 128,
      '5xl': 256,
    }[props.size];
  });

  /**
   * Get the path of the icon.
   */
  const path = computed(() => {
    // if the icon has - in the name, delete them and capitalize the next letter
    // e.g. mdi-arrow-right-bold => mdiArrowRightBold
    // and add mdi prefix
    const iconName = props.icon.startsWith('mdi')
      ? props.icon
      : `mdi${props.icon
          .split('-')
          .map((word) => {
            return word.charAt(0).toUpperCase() + word.slice(1);
          })
          .join('')}`;
    let path = mdiIcons[iconName];
    if (!path) {
      console.error(`Icon ${props.icon} not found.`);
      path = mdiIcons['mdiAlert'];
    }
    return path;
  });
</script>

<style>
  .icon {
    @apply flex w-fit items-center justify-center;
  }
</style>
