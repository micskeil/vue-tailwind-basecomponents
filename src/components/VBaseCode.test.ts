import BaseCode from './BaseCode.vue';
import BaseIcon from './BaseIcon.vue';
import { mount } from '@vue/test-utils';
import { describe, it, expect, vi } from 'vitest';

// mock the clipboard API
const clipboard = {
  ...navigator.clipboard,
  writeText: vi.fn(),
};
Object.defineProperty(window.navigator, 'clipboard', { value: clipboard });

describe('CodeHighlighter', () => {
  it('renders the given code with proper syntax highlighting', () => {
    const wrapper = mount(BaseCode, {
      propsData: {
        code: 'console.log("Hello World!");',
        language: 'javascript',
      },
      global: {
        components: {
          BaseIcon,
        },
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('copies the code when copy icon is clicked', async () => {
    const wrapper = mount(BaseCode, {
      propsData: {
        code: 'console.log("Hello World!");',
        language: 'javascript',
      },
      global: {
        components: {
          BaseIcon,
        },
      },
    });

    const copyIcon = wrapper.findComponent({ name: 'BaseIcon' });
    await copyIcon.trigger('click');

    expect(window.navigator.clipboard.writeText).toHaveBeenCalledWith(
      'console.log("Hello World!");'
    );
    expect(wrapper.find('.copy__confirmation').exists()).toBe(true);
  });
});
