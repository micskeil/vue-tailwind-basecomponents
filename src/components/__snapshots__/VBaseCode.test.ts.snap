// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodeHighlighter > renders the given code with proper syntax highlighting 1`] = `
"<div data-v-3441152e=\\"\\" data-v-d41bccca=\\"\\" class=\\"vb-input-container vb-code-container\\">
  <div data-v-d41bccca=\\"\\" class=\\"vb-code-tools\\">
    <div data-v-d41bccca=\\"\\" class=\\"copy\\">
      <!--v-if--><span data-v-d41bccca=\\"\\" class=\\"vb-icon copy__icon\\" role=\\"img\\"><svg class=\\"vb-icon__native fill-primary vb-icon-content-copy\\" viewBox=\\"0 0 24 24\\" width=\\"16\\"><path d=\\"M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z\\"></path></svg></span>
    </div>
  </div><pre data-v-d41bccca=\\"\\" class=\\"vb-code\\">      <!-- eslint-disable-next-line vue/no-v-html -->
      <code data-v-d41bccca=\\"\\" class=\\"language-javascript\\">console<span class=\\"token punctuation\\">.</span><span class=\\"token function\\">log</span><span class=\\"token punctuation\\">(</span><span class=\\"token string\\">\\"Hello World!\\"</span><span class=\\"token punctuation\\">)</span><span class=\\"token punctuation\\">;</span></code>
    </pre>
  <div data-v-d41bccca=\\"\\" class=\\"code-language\\">javascript</div>
</div>"
`;
